### ISSUE TITLE
WSO2 attempts to spawn multiple FIX sessions to same target under heavy initial load

### ISSUE THREAD

**Issue by PetteriPertola** [wso2/product-ei#3211]
**Description:**
We are doing MQ (JMS queue) to FIX message transformation. In an error scenario we may have had our WSO2 integrator (v. 6.4.0) instance down for some minutes. This means there are a large number of messages waiting on the inbound MQ. As soon as we start up, these are immediately consumed by the WSO2 proxy service consuming the queue (**good**), but given there are hundreds of messages in the queue, WSO2 then initializes QuickfixJ initiator/client multiple times (**bad**), or at least that is what it looks like in our logs. 

This results in the following behaviour: three QuickfixJ client instances attempt to logon using the same FIX session to the same target exactly at the same time, as seen by the logs here. The first one successfully logs on, but the other two instances continue attempting to logon at ~0.5 second intervals:
20190213-11:47:**45.322**: 8=FIXT.1.1 9=86 35=A 34=1 49=INITIATOR 52=20190213-11:47:45.321 56=TARGETCOMPID 98=0 108=30 141=Y 1137=9 10=176 
20190213-11:47:**45.325**: 8=FIXT.1.1 9=86 35=A 34=1 49=INITIATOR 52=20190213-11:47:45.325 56=TARGETCOMPID 98=0 108=30 141=Y 1137=9 10=180 
20190213-11:47:**45.326**: 8=FIXT.1.1 9=86 35=A 34=1 49=INITIATOR 52=20190213-11:47:45.325 56=TARGETCOMPID 98=0 108=30 141=Y 1137=9 10=180 
**35=A logon response below**
20190213-11:47:45.333: 8=FIXT.1.1 9=86 35=A 34=1 49=TARGETCOMPID 52=20190213-11:47:45.330 56=INITIATOR 98=0 108=30 141=Y 1137=9 10=176 
**Continued logon attempts afterwards:**
20190213-11:47:49.212: 8=FIXT.1.1 9=86 35=A 34=1 49=INITIATOR 52=20190213-11:47:49.211 56=TARGETCOMPID 98=0 108=30 141=Y 1137=9 10=178 
20190213-11:48:15.316: 8=FIXT.1.1 9=86 35=A 34=1 49=INITIATOR 52=20190213-11:48:15.315 56=TARGETCOMPID 98=0 108=30 141=Y 1137=9 10=177 

This is of course not allowed on FIX protocol level, so in the other application there are lots of errors as a result, and the FIX log is filled with logon messages. 

The only recovery is to restart the WSO2 instance completely to force a clean logon.

This is a clear concurrency issue. The FIX sender/client should be a singleton instance!

**Suggested Labels:**
FIX integration, concurrency issue

**Suggested Assignees:**

**Affected Product Version:**
Lates enterprise integrator v. 6.4.0

**OS, DB, other environment details and versions:**    
Linux, reproducible in all OSs.

**Steps to reproduce:**
1. Put a large (>100) quantity of messages to a JMS queue. 
2. Create a proxy service that listens to the JMS queue and sends a message to a FIX endpoint after basic message transformation. 
3. Deploy to a live WSO2 instance and observe multiple logon attempts to target FIX server. 

**Related Issues:**

**Comment by arunans23** [comment]
The FixTransportSender uses a Singleton to create a FIXsession. Also the login attempt happens at the start of the WSO2 Instance regardless of the messages available in MQ.
Your problem might occur due to different problem. It would be helpful if you can share your synapse configuration (an abstract one if confidential), to provide more overview on this matter.

**Comment by PetteriPertola** [comment]
Sure. Here is the proxy service, and the Sequence it calls:

```
<?xml version="1.0" encoding="UTF-8"?>
<proxy name="REQ_IN" startOnLoad="true" statistics="enable" transports="jms" xmlns="http://ws.apache.org/ns/synapse">
    <target>
        <inSequence>
            <property name="OUT_ONLY" scope="default" type="STRING" value="true"/>
            <property name="transport.fix.ServiceName" scope="axis2-client" type="STRING" value="FIXProxy"/>
            <sequence key="conf:/sequences/SEQ_MarketDataRequest.xml"/>
        </inSequence>
        <outSequence/>
        <faultSequence/>
    </target>
    <parameter name="transport.jms.Destination">REQ_JNDI_IN</parameter>
    <parameter name="transport.jms.ContentType">
        <rules xmlns="">
            <jmsProperty>contentType</jmsProperty>
            <default>application/xml</default>
        </rules>
    </parameter>
    <parameter name="transport.jms.DefaultReplyDestination">REQ_JNDI_OUT</parameter>
</proxy>
```

The **fixEndPointUri** in the below snippet is created as:
fix://hostname:port?BeginString=FIXT.1.1&SenderCompID=SENDER&TargetCompID=TARGET

```
<?xml version="1.0" encoding="UTF-8"?>
<sequence name="SEQ_MarketDataRequest" trace="disable" xmlns="http://ws.apache.org/ns/synapse">
    <payloadFactory media-type="xml">
        <format>
            <message xmlns="">
                <header>
                    <field id="35">V</field>
                </header>
                <body>
                    <field id="262">$1</field>
                    <field id="263">$2</field>
                    <groups id="146">
                        <group>
                            <field id="55">$3</field>
                            <field id="29998">$4</field>
                            <field id="64">SPOT</field>
                        </group>
                    </groups>
                </body>
                <trailer/>
            </message>
        </format>
        <args>
            <arg evaluator="xml" expression="//MarketDataRequest/MDReqID" xmlns:m0="http://services.samples.xsd"/>
            <arg evaluator="xml" expression="get-property('type')" xmlns:m0="http://services.samples.xsd"/>
            <arg evaluator="xml" expression="//MarketDataRequest/Symbol" xmlns:m0="http://services.samples.xsd"/>
            <arg evaluator="xml" expression="//MarketDataRequest/BusinessUnit" xmlns:m0="http://services.samples.xsd"/>
            <arg evaluator="xml" expression="//MarketDataRequest/SettlDate" xmlns:m0="http://services.samples.xsd"/>
        </args>
    </payloadFactory>
    <header expression="get-property('fixEndpointUri')" name="To" scope="default"/>
    <call>
        <endpoint template="FIX_template"/>
    </call>
</sequence>
```

```
<?xml version="1.0" encoding="UTF-8"?>
<template name="FIX_template" xmlns="http://ws.apache.org/ns/synapse">
    <endpoint name="$name">
        <default/>
        <property name="transport.fix.InitiatorMessageStore" scope="transport" value="file"/>
    </endpoint>
</template>
```

Of course this is the MQ -> FIX incoming service.

The FIX proxy service for FIX -> MQ is created as follows:
```
<?xml version="1.0" encoding="UTF-8"?>
<proxy name="FIXProxy" startOnLoad="true" statistics="enable" transports="fix" xmlns="http://ws.apache.org/ns/synapse">
    <target>
        <endpoint key="conf:/endpoints/REQ_JNDI_OUT.xml"/>
        <inSequence>
            <property name="OUT_ONLY" scope="default" type="STRING" value="true"/>
            <filter regex="[Y]" source="//field[@id='35']">
                <then>
                    <log level="full">
                        <property name="MESSAGE" value="MarketDataRequestReject detected"/>
                    </log>
                    <xslt key="MarketDataRequestRejectXSLT"/>
                </then>
                <else>
                    <log level="full">
                        <property name="MESSAGE" value="MarketDataSnapshotFullRefresh detected"/>
                    </log>
                    <xslt key="MarketDataRefreshFullTransformXSLT"/>
                </else>
            </filter>
        </inSequence>
        <outSequence/>
        <faultSequence/>
    </target>
    <parameter name="transport.fix.InitiatorConfigURL">file://location/quickfix.cfg</parameter>
    <parameter name="transport.fix.InitiatorMessageStore">file</parameter>
    <parameter name="transport.jms.ContentType">
        <rules xmlns="">
            <jmsProperty>contentType</jmsProperty>
            <default>application/xml</default>
        </rules>
    </parameter>
    <parameter name="transport.fix.InitiatorLogFactory">file</parameter>
</proxy>
```

**Comment by PetteriPertola** [comment]
Hi, I have more information on this now.

I can see via JMX (jconsole) that it spawns one too many initiators, which then causes this issue. I've attached screenshot below.

In normal operation, we have a total of six initiators to different FIX services - which you can see on server #2 on the right side. Server #1 on the left side shows a total of SEVEN initiators on the QuickfixJ library. That means you have created one session too many which looks like a concurrency issue inside the ei product somewhere. 

![image](https://user-images.githubusercontent.com/7345823/58171391-2dfced80-7c96-11e9-8cf2-ecdf4f2fc202.png)

**Comment by arunans23** [comment]
@PetteriPertola Thank you for your guidance and patience. The issue here is that, when there are lot of jms messages pending in the queue to consume, a set of jms worker threads are created to poll the messages from the queue. Hence when you start the integrator instance, these threads are individually trying to [create the FIXInitiator session](https://github.com/wso2/wso2-synapse/blob/v2.1.7-wso2v80/modules/transports/optional/fix/src/main/java/org/apache/synapse/transport/fix/FIXTransportSender.java#L306).

**Comment by PetteriPertola** [comment]
Thank you. I thought as much. Is it possible to provide a patch version for this or when can I put your fix to test?

**Comment by arunans23** [comment]
@PetteriPertola We will be testing the fix and merging it to the master branch very soon.  It will be definitely available in the next milestone / update release of WSO2 EI.

**Comment by PetteriPertola** [comment]
Thank you - do you have any indication of when that might be available -
even a ballpark would be nice? This summer? Next year?

Regards,
Petteri Pertola
Willy Brandts Vej 11, 4. mf
2450 Copenhagen SV
DENMARK

Mob: (+45) 24 46 34 21

**Comment by arunans23** [comment]
Definitely this year. You can expect a release this summer. 😄

**Comment by PetteriPertola** [comment]
Great news!
Petteri Pertola
Willy Brandts Vej 11, 4. mf
2450 Copenhagen SV
DENMARK

Mob: (+45) 24 46 34 21

**Comment by PetteriPertola** [comment]
Is this included in version 6.5.0?

**Comment by arunans23** [comment]
No. This fix was tested and merged after the release of 6.5.0. So unfortunately this is not available in the 6.5.0 release. As I mentioned above, we will be including it in a release which will be available this summer.

**Comment by PetteriPertola** [comment]
Thank you - will I get a notification here when it is released?

Petteri Pertola
Willy Brandts Vej 11, 4. mf
2450 Copenhagen SV
DENMARK

Mob: (+45) 24 46 34 21

**Comment by arunans23** [comment]
I'll post a reply to this thread. Or else you can also subscribe to our dev mailing lists. https://wso2.com/mail/ In the announcement mail, you can get all the release announcements.

**Comment by PetteriPertola** [comment]
Hi @arunans23 , what is the status of this being released? It is affecting our solution on a weekly basis.

**Comment by arunans23** [comment]
@PetteriPertola,
We are working on a release right now. It will be released next mid October. (EI 6.6.0). We will release an alpha version later this week. Please check our dev mails for more details.

**Comment by PetteriPertola** [comment]
Thank you - I have subscribed to the mailing lists so waiting eagerly for the release :-)

**Comment by PetteriPertola** [comment]
Hi @arunans23: any update on when this is released? It's now mid November :-)

**Comment by arunans23** [comment]
@PetteriPertola,
Release Candidate pack for EI 6.6.0 is available now. We will do the GA release once the voting is closed.

**Comment by PetteriPertola** [comment]
Good news. Do you have a ballpark estimate when the 'real' release is officially out?

**Comment by arunans23** [comment]
@PetteriPertola EI 6.6.0 now is officially released. You can download it in here https://wso2.com/enterprise-integrator/6.6.0

**Comment by PetteriPertola** [comment]
Hi @arunans23 ,
Did you guys test the FIX transport as a part of the 6.6.0 release? The FIX transport is now broken when you enable it due to the upgrade to log4j2 in 6.6.0 from 6.5.0.
If you keep the current version of Quickfixj that you ship (1.5.3), then you get an error relating to log4j1:
TID: [-1234] [] [2020-01-14 12:31:51,480] ERROR {org.apache.synapse.transport.fix.FIXTransportListener} - Unexpected error when configuring service FIXProxy_RFQ for the FIX tran
sport. It will be disabled for this transport and marked as faulty. **java.lang.NoClassDefFoundError: org/apache/log4j/Level**
**Caused by: java.lang.ClassNotFoundException: org.apache.log4j.Level cannot be found by quickfixj_1.5.3.wso2v1**

If i manually go in and replace the QuickfixJ plugin 1.5.3 with a newer version that supports log4j2, such as Quickfixj 2.1.0, I get a JMX export error since this part has changed in newer version of QuickfixJ:
TID: [-1234] [] [2020-01-14 12:35:34,587] ERROR {org.apache.synapse.transport.fix.FIXTransportListener} - Unexpected error when configuring service FIXProxy_RFQ for the FIX tran
sport. It will be disabled for this transport and marked as faulty. java.lang.NoSuchMethodError: org.quickfixj.jmx.JmxExporter.export(Lquickfix/Connector;)V

**Comment by arunans23** [comment]
@PetteriPertola 
Since we have moved from log4j to log4j2 we have removed all the related log4j jars from the runtime. You can resolve the above issue by adding a log4j jar to the lib folder. (For example you can try out with log4j-1.2.13.jar which is available in 6.5.0 pack) You can continue to use the existing quickfixj jar.

**Comment by PetteriPertola** [comment]
Hi, yes this works.
You should update your documentation regarding it for FIX Transport :)
